main.go

```diff
@@ -3,74 +3,71 @@ package main
 import (
 	"bytes"
 	"context"
-	"fmt"
+	"flag"
+	"github.com/Riskified/worker_runner/internal/healthcheck"
+	log "github.com/sirupsen/logrus"
 	"io"
-	"log"
-	"net/http"
 	"os"
 	"os/exec"
 	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
-	"sync"
 	"syscall"
 )
 
-func healthcheck(w http.ResponseWriter, req *http.Request) {
-	log.Print("got healthcheck")
-	fmt.Fprintf(w, "ok\n")
-}
 
-func startWorker(stopChan <-chan struct{}, wg *sync.WaitGroup, command string, args ...string) {
+
+func startWorker(stopChan <-chan struct{}, command string, args ...string) (doneChan <-chan struct{}) {
 	cmd := exec.Command(command, args...)
 
 	var stdoutBuf, stderrBuf bytes.Buffer
 	cmd.Stdout = io.MultiWriter(os.Stdout, &stdoutBuf)
 	cmd.Stderr = io.MultiWriter(os.Stderr, &stderrBuf)
 
+	done := make(chan struct{})
+
 	go func() {
-		err := cmd.Run()
+		err := cmd.Start()
 		if err != nil {
-			log.Fatalf("cmd.Run() failed with %s\n", err)
+			log.WithField("error", err).Error("Failed starting command")
+			close(done)
 		}
+		log.Info("worker started")
+
+		go func() {
+			<-stopChan
+			log.Info("send worker SIGTERM")
+			cmd.Process.Signal(syscall.SIGTERM)
+		}()
+
 		cmd.Wait()
-		wg.Done()
-		log.Print("worker ended")
+		close(done)
 	}()
 
-	<-stopChan
-	log.Print("send worker SIGTERM")
-	cmd.Process.Signal(syscall.SIGTERM)
-}
 
-func startServerAsync() *http.Server{
-	srv := &http.Server{Addr: ":8000"}
-	http.HandleFunc("/healthcheck", healthcheck)
-	go func() {
-	if err := srv.ListenAndServe(); err != nil {
-		log.Fatalf("ListenAndServe(): %v", err)
-	}
-	}()
 
-	return srv
+	return done
 }
 
 func main() {
-	if len(os.Args) == 1 {
+	var port int
+	flag.IntVar(&port, "port", 8080, "port for healthcheck")
+
+	flag.Parse()
+	if len(flag.Args()) == 0 {
 		log.Fatal("command not found. usages: worker_runner COMMANDS ARGS")
 	}
-	stopChan := signals.SetupSignalHandler()
 
-	wg := &sync.WaitGroup{}
+	log.Info(flag.Args())
 
-	log.Print("start worker")
-	wg.Add(1)
-	go startWorker(stopChan, wg, os.Args[1], os.Args[2:]...)
+	log.Info("start worker")
+	done := startWorker(signals.SetupSignalHandler(), flag.Args()[0], flag.Args()[1:]...)
 
-	log.Print("start server")
-	srv := startServerAsync()
+	log.Info("start server")
+	srv := healthcheck.StartServerAsync(port)
 
-	wg.Wait()
-	log.Print("shutdown server")
+	<- done
+	log.Info("worker ended")
```

Comment:
Note that at this point if could be that the Worker is no longer running (it gets a signal from the Wrapper) but the health check is still running (hence potentially return wrong Healthy respond). 

In that case, if the health check is for the Wrapper ONLY health (and NOT for the Worker) then this is Ok.

Reply:
One or two incorrect healthcheck in the shoutdown process are OK, the prcess will die soon